### PR TITLE
P10->P11 migration into Rust block state

### PR DESCRIPTION
**Pull request by allanbrondum**
## Purpose

Implement migration of the Rust block state from P10 to P11 on the Haskell side.

## Changes

The P10 PLT state is loaded by Rust and then migrated to P11 by Rust also.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs [hunk 1]
```diff
@@ -193,15 +194,21 @@ emptyProtocolLevelTokens =
 -- * 'PLTStateNone': No PLT state
 -- * 'PLTStateV0': Managed in Haskell
 -- * 'PLTStateV1': Managed in Rust
-data ProtocolLevelTokensForStateVersion (pv :: ProtocolVersion) where
-    ProtocolLevelTokensNone :: (PltStateVersionFor pv ~ 'PLTStateNone) => ProtocolLevelTokensForStateVersion pv
-    ProtocolLevelTokensV0 :: (PltStateVersionFor pv ~ 'PLTStateV0) => (HashedBufferedRef' ProtocolLevelTokensHash ProtocolLevelTokens) -> ProtocolLevelTokensForStateVersion pv
-    ProtocolLevelTokensV1 :: (PltStateVersionFor pv ~ 'PLTStateV1) => RustBS.ForeignPLTBlockStatePtr pv -> ProtocolLevelTokensForStateVersion pv
+data ProtocolLevelTokensForPV (pv :: ProtocolVersion) where
```
Comment (allanbrondum) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs:196]: Just renamed and reformatted

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs [hunk 2]
```diff
@@ -479,32 +486,41 @@ migrateProtocolLevelTokens ProtocolLevelTokens{..} = do
...
+            -- migrate the newly loaded state in the Rust block state implementation.
+            let oldBlobRefMaybe = getHBRRefIfBlobbed oldPLTsRef
+            let oldBlobRef =
+                    maybe
+                        (error "Haskell maintained PLT state to migrate from does not have blob reference set")
+                        id
+                        oldBlobRefMaybe
+            (oldState :: ForeignPLTBlockStatePtr (MPV m)) <- lift $ loadDirect (coerce oldBlobRef)
+            ProtocolLevelTokensV1 <$> RustBS.migrate oldState
+migrateProtocolLevelTokensForPV migration (ProtocolLevelTokensV1 oldState) =
```
Comment (allanbrondum) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs]: The actual migration logic

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs [hunk 3]
```diff
@@ -479,32 +486,41 @@ migrateProtocolLevelTokens ProtocolLevelTokens{..} = do
...
-migrateProtocolLevelTokensForStateVersion migration (ProtocolLevelTokensV1 oldState) =
+        SPLTStateV1 -> do
+            -- Migrate from Haskell to Rust, by reloading the PLT state from the blob store and into Rust and then
+            -- migrate the newly loaded state in the Rust block state implementation.
+            let oldBlobRefMaybe = getHBRRefIfBlobbed oldPLTsRef
+            let oldBlobRef =
+                    maybe
+                        (error "Haskell maintained PLT state to migrate from does not have blob reference set")
+                        id
+                        oldBlobRefMaybe
```
Comment (td202) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs:520]:
```suggestion
            (_, oldBlobRef) <- lift $ refFlush oldPLTsRef
```

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlobStore.hs [hunk 4]
```diff
@@ -1073,6 +1074,14 @@ makeBufferedRef v = liftIO $ do
...
 blobRefToBufferedRef = BRBlobbed
 
+-- | Get the 'BlobRef' if the value in the reference is stored in the blob store. Else
+-- 'Nothing' is returned.
+getBRRefIfBlobbed ::
+    BufferedRef a -> Maybe (BlobRef a)
+getBRRefIfBlobbed (BRBlobbed ref) = Just ref
+getBRRefIfBlobbed (BRMemory _ _) = Nothing
+getBRRefIfBlobbed (BRBoth ref _) = Just ref
+
```
Comment (td202) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlobStore.hs:1084]: Prefer using `refFlush`

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlobStore.hs [hunk 5]
```diff
@@ -1690,6 +1699,12 @@ makeHashedBufferedRef val = do
     hashRef <- liftIO $ newIORef Null
     return $ HashedBufferedRef br hashRef
 
+-- | Get the 'BlobRef' if value in the reference is stored in the blob store. Else
+-- 'Nothing' is returned.
+getHBRRefIfBlobbed ::
+    HashedBufferedRef' h a -> Maybe (BlobRef a)
+getHBRRefIfBlobbed (HashedBufferedRef br _) = getBRRefIfBlobbed br
+
```
Comment (td202) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlobStore.hs:1707]: Prefer using `refFlush`.

**Review by td202 — CHANGES_REQUESTED** [review]

File: concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs
Comment (allanbrondum) [inline concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs:1]: Smoketest of Rust block state ffi functions

File: concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    (loaded :: ForeignPLTBlockStatePtr 'P11) <- loadRef =<< storeRef state
+    (hashAfter :: ProtocolLevelTokensHash) <- getHashM loaded
+    liftIO $ assertEqual "Hash should be preserved across store/load" hashBefore hashAfter
+    -- Cache state
+    (_cached :: ForeignPLTBlockStatePtr 'P11) <- cache state
+    return ()
+
+-- | Test migrate state.
+testMigrate :: Assertion
+testMigrate = do
```
Comment (soerenbf) [inline concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs:41]:
```suggestion
testMigrateEmpty :: Assertion
testMigrateEmpty = do
```

File: concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        -- Migrate
+        targetStore <- liftIO newMemBlobStore
+        flip runMemBlobStoreT targetStore $ do
+            (_migrated :: ForeignPLTBlockStatePtr 'P11) <- migrate state
+            return ()
+
+tests :: Spec
+tests = describe "GlobalStateTests.RustPLTBlockState" $ do
+    it "storeLoadHashCache" testStoreLoadHashCache
+    it "migrate" testMigrate
```
Comment (soerenbf) [inline concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs]:
```suggestion
    it "migrate empty PLT state p10->p11" testMigrateEmpty
```

to be more descriptive?

File: concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs
Comment (soerenbf) [inline concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs:1]: Would it make sense to add a test with a non-empty haskell PLT state and verify that it migrates to the same non-empty corresponding rust state?

**Review by soerenbf — APPROVED** [review]
Assuming @td202's comments are taken into account

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens/RustPLTBlockState.hs [hunk 8]
```diff
@@ -206,13 +206,15 @@ migrate ::
 migrate currentState = do
     oldLoadCallback <- fst <$> lift BlobStore.getCallbacks
     newStoreCallback <- snd <$> BlobStore.getCallbacks
+    newLoadCallback <- fst <$> BlobStore.getCallbacks
```
Comment (limemloh) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens/RustPLTBlockState.hs]: Nitpick (non-blocking)
```suggestion
    (newLoadCallback, newStoreCallback) <- BlobStore.getCallbacks
```

File: concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs (hunk 7 above)
Comment (limemloh) [reply concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs]: I suggest the description captures the value of the test, rather than summarizing what it does.
For this one it seems we don't have any assertions, so I guess the point is to check it does not fail

**Review by limemloh — APPROVED** [review]

File: concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs
Comment (allanbrondum) [reply concordium-consensus/tests/globalstate/GlobalStateTests/RustPLTBlockState.hs:1]: We can't really assert on the migrated state. Only at scheduler level through queries. I think we have to postpone implementing a test suite at that level for now. Current test is mostly a smoketest of the FFI interface.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs (hunk 3 above)
Comment (allanbrondum) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs:520]: We discussed it and decides it was better to fail hard than silently, if the state is not already on disk.
